synth.py

```diff
@@ -17,18 +17,29 @@ def sine_wave(duration, frequency, ampl=1.0, samplerate=SAMPLERATE):
     return (0.5 * ampl) * np.sin(x * frequency * np.pi * 2)
 
 
-def envelope(attack_time, decay_time, sustain_level, release_time, frames):
+def envelope(attack_time, decay_time, sustain_level, release_time, frames, samplerate=SAMPLERATE, msecs=False):
     assert isinstance(frames, int)
-    attack_frames = int(frames * attack_time)
-    decay_frames = int(frames * decay_time)
-    release_frames = int(frames * release_time)
-    sustain_frames = frames - attack_frames - decay_frames - release_frames
+
+    if msecs is True:
+        attack_frames = np.clip(int(attack_time / 1000 * samplerate), 0, None)
+        decay_frames = np.clip(int(decay_time / 1000 * samplerate), 0, None)
+        release_frames = np.clip(int(release_time / 1000 * samplerate), 0, None)
+        sustain_frames = 0
+        padding_frames = np.clip(frames - attack_frames - decay_frames - release_frames, 0, None)
+    else:
+        attack_frames = int(frames * attack_time)
+        decay_frames = int(frames * decay_time)
+        release_frames = int(frames * release_time)
+        sustain_frames = frames - attack_frames - decay_frames - release_frames
+        padding_frames = 0
```

Comment:
1) why only msecs are clipped?
2) why padding and sustain are swapped compared to above?
3) if you divide the attack/decay/release time by 1000 before the if/else, you might be able to get rid of the duplication

Reply:
Fair enough, it makes sense to clip outside the if/else even if it's only really problematic if you use the milliseconds mode.
Regarding 2 and 3, the two modes of operation works in a different way. In the first case there's no need to add silence frames because all the ADSR times are stretched according to the duration of the note; that's not so in the new algorithm. Also, the division by 1000 cannot be moved outside, because only in the new mode of operation you are specifying milliseconds.